MPR#7574: hightlight differences in type error messages

Octachron opened the pull request:
This PR proposes to make the differences in type error messages more visible
by using both color (beware that the svg rendering is a little whimsical)

<img alt="type error with an optional argument"  src="http://polychoron.fr/pr1481/options.svg"/>
<img alt="type error with wrong types for function argument "  src="http://polychoron.fr/pr1481/arrow.svg"/> 
<img alt="type error with an object type"  src="http://polychoron.fr/pr1481/object_0.svg"/>


and collapsing parts of the types that are not relevant


<img alt="type error with a large object types where relevant part are elided with ..."  src="http://polychoron.fr/pr1481/object_1.svg"/>
<img alt="type error with a polymorphic variant definition"  src="http://polychoron.fr/pr1481/polyvar.svg"/>
<img alt="type error with a conflict between implementation and interface for a module "  src="http://polychoron.fr/pr1481/modules.svg"/>

This PR works by duplicating the outcometree in a plain and decorated version. Most of the producers of outcometree can then still produce a plain outcometree. (see 8ccd645 ).

Nevertheless, the outcometree printers defined in `typing/oprint.ml` consumes the decorated version of the outcome tree. The standard printers defined in `printtyp.ml` have been updated to lift the plain outcometree to a plain decorated outcometree and print it. Type errors that only involve one type still
uses these printers (see bfa8427 ).

Contrarily, a new module has been introduced to compare and collapse two outcome trees. Difference are highligted, and the trees are collapsed in order to fit a size limit customizable through a new `-error-size` option (see bbb2c88 ).

As a complement to color highlighting, a new option adds the possibility to delimit difference between textual marker "!(' ")", both as an accessibility attempt (but @shindere was sceptical) and for testing purpose.

More examples are available in `testsuite/test/typing-error-messages`, however they are rendered with a very small (=2) error size limit to test the collapsing feature.

Octachron committed:
highlightable and collapsable outcometree

Octachron committed:
make Oprint work with the decorated outcome tree

Octachron committed:
Module for outcome tree difference computation

Octachron committed:
use difference higlighting in unification error

Octachron committed:
Add -error-size and -diff-markers option

Octachron committed:
Add tests for type diff in error messages

Octachron committed:
Manual documentation

Octachron committed:
Changes

v-gb commented:
Can you clarify whether the diff is syntactic or semantic, ie whether the diff of `int -> foo` and `float -> alias_of_foo` would highlight `foo` and `alias_of_foo` or not?

Drup commented:
@sliquister The diff is done on the structure representing type expressions (more precisely, the one that is used specifically for user output, not the one used for typechecking). It's a great approach and might allow really nice work down the road. There has been various projects trying to recover back these diffs outside the compiler, this is much cleaner.

The nice thing about this work is that, now that we have nice structures containing the diffs, we can add lot's of heuristics to make the result nicer. For example, showing small modules/objects where only one field is different leads to repeated diffs that don't bring further information. That's for the next PR though :p

@Octachron I started looking at it, and everything is pretty much trivial ... except `typing/difftree.ml`. Could you elaborate a bit on which algorithm you use here? There are comments about details, but not about the general idea.

There are other trivial remark: You take great care of making `Outcometree` generic. You should just write the elaborator as a generic functor that takes two extensions, a map, and produce a mapper (*cough* visitors *cough*). `Decorate` is then a trivial functor application. Speaking of which, I would prefer more uniform names (Highlightable/Highlighted, something like that).

Octachron commented:
@sliquister , the difference is not completely semantic currently and there would be false positive in the highlighted difference currently. For instance, `alias_of_foo` will be highliged when comparing `int -> foo` and `float -> alias_of_foo`. This could be fixed however with enough time. Moreover, for this specific example, I would say that is the job of `-short-path` to compute canonical best path for types.

@Drup, thanks for the remarks!

For the second point, do you mean having `Highlighted/Plain` rather than `On/Off` or are you speaking of having a `Highlighted` and `Plain` modules rather than a `Decorate` module?

For the first point, let me try to collect my thoughts before putting them in comments:

In term of algorithm for difftree.ml, the basic hypothesis is that edit distances on outcome tree will not be that interesting, since far too often the root cause of errors will be too removed from the generated outcome tree. The idea is therefore to focus on small differences between large types; leaving the more complex situation to the unification trace and human intellect.

In this context of small differences between large types, printing integrally very large types is too noisy. Therefore, the algorithm implemented try to limit the size of the printed outcome tree. This is done by working on tree generators rather than the tree themselves.

This is the motivation behind the `gen` type

```OCaml
type ('a,'b) gen = {
  gen: int -> 'a; (** [gen fuel x] generates a representation of the underlying
                      element with size [fuel] *)
  size:'b; (** size represents the maximal size in term of node or leaves of
               the generated elements *)
}
```

The size type is furthermore splits in two components:

```
  type t = {
    (** size of elements that should be printed in priority *)
    primary: int;

    (** size of elements that may be printed if there is some spare space *)
    secondary: int
  }
```

The primary component is currently used to account for the number of elements that differs between the two compared trees, whereas the secondary tree measure the remaining elements in the generated trees.

This graded size is used when constructing generators from theirs subtrees. The parent generator distributes its size budget (currently misnamed fuel in the code) to its children by considering first the primary size requirements and then distribute any remaining size budget by looking at the secondary size requests.

The distribution algorithm currently used during each phase is `Applist.rising_tide`: it first distributes size budget to small elements that can be completely expanded with the current budget then distributes greedily the remaining budget from left to right.

Any element which was not attributed any size budget is then replaced by an ellipsis (of length 1) and ellipsises are then concatenated. 

The implementation of this distribution algorithm comes in four flavors:

* `Applist.mkdiff` and `AppList.(<*>)` implement, it for nodes with fixed numbers of
children (i.e. all constructor in the outcome tree types).

The three other flavors are here to deal with variable length children list and possible mismatch between their shape. All of them represent missing elements in the children list by holes implemented as zero length ellipsis.

* `List_combinators.list` handles simple ordered list.

* `List_combinators.keyed_list` is used for polymorphic variants and objects, where lists are ordered by a key parameter with an explicit ordering of keys.

* `List_combinators.module_like` is used for modules, where the specific order of the
elements should be preserved by the printing but elements should be compared using
their keys.

The highlighting side of the algorithm is more straightforward and corresponds mostly to a simple `map2` on the outcome tree comparing the current subtree and highlighting them if they are different.

Like I said to @sliquister , the comparison here is still approximate and their are some false positive that could be eliminated with more work and time.

gasche commented:
Personally I think it's already a tremendous improvement, and that we should consider merging this even with the current limitation (I agree that letting -short-path deal with short canonical names is a reasonable design choice.)

v-gb commented:
@gasche short-paths normalizes paths into paths. It doesn't help with `type t = int * int`, and an error with `t` on one side and `int * int` on the other (which assumes the typer hasn't unified them).

@Octachron false positives looks like they should be rather frequent, is that right? It seems to me that many type errors, in particular module type errors, would show red that are not errors. For instance the following example would show red in t and x and y when only y is wrong:

```ocaml
module M1 : sig
  type t
  val x : t
  val y : t
end = struct
  type t = int * int
  let x = 1, 1
  let y = 1., 1
end
```

You also have false negatives, like this error showing no red at all:

```ocaml
module M : sig
  val x : 'a
end = struct
  let x = 1
end
```

Octachron committed:
difftree: distinguish unification/inclusion error

Octachron committed:
difftree: abstracted type in module signature

Octachron commented:
False positives like your first example will definitely happen. However, in this case, the rest of the error message details the real error: 

> Values do not match: val y : float * float is not included in val y : t

False positive on unification errors are also possible but a bit more subtle:

```OCaml
type t = int * int
let g (x: int * int) = ()
let f (g: t -> t) = ()  
;; f g
```

yields the right highlighting

> This expression has type t -> !(unit) but an expression was expected of type t -> !(t)

However going further, 
```OCaml
type t = int * int
type 'a s = 'a * 'a
let g (x: int s) = ()
let f (g: t -> t) = ()  
;; f g
```

yields a false positive

> This expression has type !(int s) -> !(unit) but an expression was expected of type !(t) -> !(t)

These false positives could be eliminated by doing a three-way difference between the actual type, the partially unified type and the expected type; but I am not sure if it is worthwhile to complexify the code further at this point.

Contrarily, false negatives like 

```OCaml
module M : sig
  val x : 'a
end = struct
  let x = 1
end
```

were a bug due to having a heuristic for decreasing false positives in unification errors in presence of type variables that was not valid for inclusion errors. I have fixed this mistake.

Octachron committed:
difftree: make stitch more generic

Armael commented on driver/main_args.ml:
```diff
@@ -77,6 +77,11 @@ let mk_dllpath f =
+  "-error-size", Arg.Int i,
+  "<size>  Increase the size of error message to <size> (in arbitrary units)"
+;;
+
```
Since its in "arbitrary units", would it make sense to give some indication of what is a reasonable value for this? And/or indicate what the default value is?

Octachron replied:
This is a very good point. I added the default value in the `-help` description and a typical range in the various manuals.

lpw25 commented:
I have mixed feelings about this PR. On the one-hand it does improve the error messages and is a fairly isolated change. On the other, every example you've shown demonstrates a failure of the existing error handling that would be better fixed by improving that handling. A syntactic approach is always going to have false positives, whilst fixing the underlying error reporting would always be accurate. I'm a little worried that adding the syntactic diffs will stop people from fixing the underlying problems with error reporting since they won't arise as often -- although I don't know of anyone currently planning to fix these problems especially given the impending rewrite of the type checker.

gasche commented on utils/misc.ml:
```diff
@@ -566,12 +566,14 @@ module Color = struct
     warning = [Bold; FG Magenta];
     error = [Bold; FG Red];
     loc = [Bold];
+    difference = [Bold; FG Red];
```
In the tradition that red is reserved for hard errors, I would rather use a softer color for things that may contain false positives, such as orange (my first idea), the warning color (magenta), or just bold.

Octachron replied:
Good point, I chose magenta since I find it much better in term of constrast with red and it seems fine even for colorblind people.

gasche commented on utils/misc.ml:
```diff
@@ -584,21 +586,28 @@ module Color = struct
 
   let color_enabled = ref true
+  let marker_enabled = ref false
+  let diff_markers = ref ("!(",")")
```
What is the impact of this choice of ASCII representation for diff markers? Am I correct that they will never be shown to the user, only used by the internal Format machinery?

Octachron replied:
They are only visible when requested with the `-diff-marker` option, with the idea that such marker are friendlier to parse than terminal escape sequences (initially, I added them as an accessibility option, but they seem to be a failure on this point)

gasche reviewed (commented):
I only had a brief look, this is not a full review;

- could you rebase against trunk so that we get the expect_test changes out of the way?

- The current approach to have two separate definitions of outcometrees (one decorated, one not) gives a lot of structure, but it is also fairly verbose. Have you tried first to just add the extra structure (emphasis nodes and explicit ellipses) to plain outcometrees, and define the diff at that level?

gasche commented:
> I'm a little worried that adding the syntactic diffs will stop people from fixing the underlying problems with error reporting since they won't arise as often.

I think that's an argument that is not terribly convincing in general (I used it twice myself, once against making ocamlyacc nicer because a Merlin parser was just around the corner, and once against making the toplevel nicer because obviously everyone would soon be using `utop`, and now is as good a time as any to recognize that I appear to have been mistaken on both accounts). Given the impredictability of these sort of works, I would rather go for the short-term improvement even when there is a long-term solution down the road.

Also, in cases like this one where we worry about false positives, I think that some experience with the feature could be helpful. We could get that sort of experience by just merging in trunk and seeing what happens.

bluddy commented:
> impending rewrite of the type checker.

A rewrite? How come?

lpw25 commented:
> I think that's an argument that is not terribly convincing in general 

I basically agree that my argument here isn't convincing. Although note that it is a slightly different argument that the one you are describing. My concern is more that we will not get the bug reports about bad error messages, rather than saying that we shouldn't improve things in the short-term. My concern is that this patch is a bit like fixing code by adding
```
try
  do_something ()
with _ -> 0 (* zero is better than an uncaught exception *)
```
but as I said, I don't think this concern is particularly convincing here.

gasche commented:
@lpw25 I see; given that bad error messages are not a new issue (we've had them all along), but rather an instance of our perception changing on what kind of issues we are willing to invest efforts on, maybe we could be proactive in building a list of problematic cases and tracking them on Mantis, instead of relying on user reports?

This happens to align rather nicely with the idea of collecting a list of testcases for @Octachron's PR, containing representative instances of both clear improvements, but also dubious situations and false positives. A lot of work was already done in writing the testsuite involved in this PR, can we evaluate how complete it is in this respect, and make sure it serves as a step forward in our understanding of what should be improved?

Octachron committed:
Conversion to ocamltest

Octachron committed:
Magenta difference

Octachron committed:
More info on the size unit

Octachron commented:
@gasche , I have rebased the PR to remove the `expect_test` commit (and converted the tests to OCamltest).

> The current approach to have two separate definitions of outcometrees (one decorated, one not) gives a lot of structure, but it is also fairly verbose. Have you tried first to just add the extra structure (emphasis nodes and explicit ellipses) to plain outcometrees, and define the diff at that level?

This was my first implementation, it does work but it end up leaking the extra structure of the ouctome tree into the tree construction function of printtyp.ml (i.e. the `tree_of_*` functions) and at least one other place (that I don't remember precisely). At the end, I prefered to have the current separation with its verbose but trivial elaboration code and keep the separation of concern between building the outcome tree from `type_expr` and decorating the outcome tree. I could have been mistaken.

> , but also dubious situations and false positives. A lot of work was already done in writing the testsuite involved in this PR, can we evaluate how complete it is in this respect,

On this front, after @sliquister remarks, I have thought of more problematic case. I will try to add them to the test suite with some explanation of the reason behind these failures

v-gb commented:
I have a couple more false negatives, but there are a number of false positives, especially at the module level: you can take any signature item, and think: how can I provide a definition for that signature without using exactly the same syntax? There are probably many ways of doing that. I included examples below.

I don't agree with Leo that the error shown would be better handled some other way: it just seems more direct to have colors which show which part of the two types aren't compatible, than add text like "Label ~a only shows up on the left, and ~b only shows up on the right". The diffs don't have to be syntactic though.

What I've wanted to try but haven't had time to, to address the issue of the false positives at least for unification errors, is to use the actual functions of the typer to decide whether types differ or not when reporting error. One idea is to check compatibility of types using unification when reporting error: if `a b c` is not compatible with `d e f` then try to unify `_ c` with `_ f`. If they don't unify, then color `c` and `f`. If they unify, try to unify `_ b c` with `_ e f`. If they don't unify, if `c` is not the same constructor as `f`, then color `b c` and `e f` otherwise color `b` and `e`. If they unify, keep recursing (I think short paths would get rid of type constructors that expand to one of their parameter, which simplifies a bit the problem).

Also, at first, I thought it was suboptimal to have the colors and type expansions be partially redundant (if `int t t` and `float t t` are not compatible, then given that `int` and `float` will be highlighted, why lengthen the error by expanding `t` twice?). But now I think it's fine because it preserves the property that colors can be removed without losing information.


False negatives:
  
```ocaml
module M : sig
  val f : 'a list ref
end = struct
  let f = ref []
end
```

```ocaml
module M : sig
  val f : 'a -> 'b -> 'a * 'b
end = struct
  let f a b = b, a
end
```

False positives:

```ocaml
module M : sig
  type t
  type u
  module M : sig type t end
  val x : unit
end = struct
  type u and t
  module M = struct
    type nonrec t = t
  end
  let x = ""
end
```

```ocaml
module M : sig
  type t = [ `A ]
  type u = t
  val x : unit
end = struct
  type u = [ `A ]
  type t = u
  let x = 1
end
```

xavierleroy commented:
Two years later, what's the status on this PR?

Octachron commented:
Fixing the syntactic nature of this highlighter would require a significant amount of work. I am thus closing this issue. I will open a new PR if I eventually find the time to fix this core issue.
